### PR TITLE
Methods content

### PR DESCRIPTION
**Pull request by stubailo**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162752950%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162753556%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162756678%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162756832%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162762336%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162800743%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46914612%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46914642%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46914691%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46915044%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46922137%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46922163%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46922422%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46922423%22%2C%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-163075150%22%2C%20%22https%3A//github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277%23commitcomment-14863348%22%2C%20%22https%3A//github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277%23commitcomment-14863628%22%2C%20%22https%3A//github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277%23commitcomment-14863639%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/pull/138%23issuecomment-162752950%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22When%20is%20a%20good%20time%20to%20use%20%60onResultReceived%60%3F%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A04%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%22The%20only%20time%20I%20ever%20used%20it%20was%20when%20repeatedly%20calling%20a%20method%20that%20touched%20the%20same%20collection%2C%20and%20wanting%20to%20get%20a%20callback%20for%20a%20progress%20bar.%5Cr%5Cn%5Cr%5CnIf%20I%20recall%20correctly%2C%20what%20was%20happening%20is%20all%20the%20success%20callbacks%20for%20all%20the%20methods%20calls%20were%20firing%20at%20once%20because%20they%20were%20all%20waiting%20on%20each%20other%27s%20data.%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A09%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hmmm..%20is%20that%20worth%20mentioning%3F%20I%20can%27t%20really%20think%20of%20a%20simple%2C%20non-contrived%20example.%20Perhaps%20this%20is%20a%20feature%20people%20could%20look%20for%20in%20the%20docs.%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A20%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%22Speaking%20of%20this%20article%2C%20can%20you%20take%20a%20glance%20at%20the%20structure%3F%20It%27s%20a%20bit%20different%20than%20the%20outline%2C%20let%20me%20know%20what%20you%20think%20of%20the%20ordering.%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A20%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%27s%20looking%20really%20good.%20Definitely%20agree%20with%20the%20restructuring.%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A55%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22Outstanding%20questions%3A%5Cr%5Cn%5Cr%5Cn1.%20Should%20we%20avoid%20using%20%60%24inc%60%20in%20Todos%3F%20I%27m%20not%20sure%20how%20the%20denormalizer%20works.%5Cr%5Cn2.%20What%20to%20do%20about%20%60onResultReceived%60%3F%5Cr%5Cn3.%20How%20much%20detail%20should%20I%20write%20about%20loading%20data%20with%20Methods%3F%22%2C%20%22created_at%22%3A%20%222015-12-08T07%3A33%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%221.%20Let%27s%20discuss%20this%5Cr%5Cn2.%20I%20think%20drop%20it.%20It%27s%20in%20the%20docs%2C%20it%27s%20not%20common%20that%20you%20need%20it%2C%20and%20you%20can%20make%20passing%20reference%20to%20it%20in%20the%20spot%20where%20you%20explain%20how%20methods%20work.%5Cr%5Cn3.%20I%20think%20you%20should%20mention%20using%20a%20local%20collection%20and%20dumping%20data%20in%20there%20from%20a%20method.%20Then%20you%20can%20add%20the%20downsides%20about%20optimistic%20UI%20and%20overfetching%20on%20secondary%20access%22%2C%20%22created_at%22%3A%20%222015-12-09T01%3A11%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%204fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f%20content/methods.md%2054%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46914612%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%5C%22a%20client%5C%22%3F%20--%20I%20mean%20server-server%20methods%20are%20a%20thing.%20But%20maybe%20it%27s%20better%20to%20just%20not%20mention%20it.%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A42%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yeah%20I%20think%20if%20we%20mentioned%20it%20we%27d%20need%20to%20go%20into%20detail%20and%20explain%20the%20use%20cases%20-%20some%20sort%20of%20microservice%20situation.%20But%20we%20don%27t%20really%20do%20that%20in%20the%20guide%20so%20let%27s%20leave%20it%20for%20now.%22%2C%20%22created_at%22%3A%20%222015-12-08T07%3A27%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-08T07%3A33%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20content/methods.md%3AL1-417%22%7D%2C%20%22Pull%204fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f%20content/methods.md%2079%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46914642%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%28on%20the%20client%29%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A43%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20I%20misunderstood.%20Maybe%20%5C%22run%20the%20validation%20code%20by%20itself%2C%20without..%5C%22%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A44%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-12-08T07%3A28%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20content/methods.md%3AL1-417%22%7D%2C%20%22Commit%209b9442b5f7e40a430b58bf7a4acdeaba8320b277%20content/methods.md%2012%20366%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277%23commitcomment-14863628%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Link%20to%20http%3A//guide.meteor.com/collections.html%23local-collections%20%28I%27ll%20update%20the%20id%20later%20today%29%22%2C%20%22created_at%22%3A%20%222015-12-09T02%3A04%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20content/methods.md%3AL366%20%289b9442b%29%22%7D%2C%20%22Pull%204fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f%20content/methods.md%20410%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/pull/138%23discussion_r46915044%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Yeah%2C%20pretty%20much.%20%60%24inc%60%2C%20%60%24push%60%2C%20etc%20%2B%20any%203rd%20party%20side%20effects%22%2C%20%22created_at%22%3A%20%222015-12-08T04%3A54%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-08T07%3A33%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20content/methods.md%3AL1-417%22%7D%2C%20%22Commit%209b9442b5f7e40a430b58bf7a4acdeaba8320b277%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277%23commitcomment-14863348%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40tmeasday%20let%20me%20know%20if%20you%20think%20this%20is%20a%20good%20example.%22%2C%20%22created_at%22%3A%20%222015-12-09T01%3A36%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/448783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/stubailo%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20think%20it%27s%20a%20good%20example.%5Cr%5Cn%5Cr%5CnRe%20the%20%60onResultReceived%60%20I%20would%20at%20least%20call%20out%20that%20it%20exists%20in%20the%20%5C%22method%20call%20lifecycle%5C%22%20step%204%22%2C%20%22created_at%22%3A%20%222015-12-09T02%3A05%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/132554%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tmeasday%22%7D%7D%5D%2C%20%22title%22%3A%20%22Commit%209b9442b%20comments%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/meteor/guide/pull/138#issuecomment-162752950'>General Comment</a></b>
- <a href='https://github.com/stubailo'></a> When is a good time to use `onResultReceived`?
- <a href='https://github.com/tmeasday'></a> The only time I ever used it was when repeatedly calling a method that touched the same collection, and wanting to get a callback for a progress bar.
If I recall correctly, what was happening is all the success callbacks for all the methods calls were firing at once because they were all waiting on each other's data.
- <a href='https://github.com/stubailo'></a> Hmmm.. is that worth mentioning? I can't really think of a simple, non-contrived example. Perhaps this is a feature people could look for in the docs.
- <a href='https://github.com/stubailo'></a> Speaking of this article, can you take a glance at the structure? It's a bit different than the outline, let me know what you think of the ordering.
- <a href='https://github.com/tmeasday'></a> It's looking really good. Definitely agree with the restructuring.
- <a href='https://github.com/stubailo'></a> Outstanding questions:
1. Should we avoid using `$inc` in Todos? I'm not sure how the denormalizer works.
2. What to do about `onResultReceived`?
3. How much detail should I write about loading data with Methods?
- <a href='https://github.com/tmeasday'></a> 1. Let's discuss this
2. I think drop it. It's in the docs, it's not common that you need it, and you can make passing reference to it in the spot where you explain how methods work.
3. I think you should mention using a local collection and dumping data in there from a method. Then you can add the downsides about optimistic UI and overfetching on secondary access
- [ ] <a href='#crh-comment-Commit 9b9442b5f7e40a430b58bf7a4acdeaba8320b277'></a> &nbsp;<b><a href='https://github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277#commitcomment-14863348'>Commit 9b9442b comments</a></b>
- <a href='https://github.com/stubailo'></a> @tmeasday let me know if you think this is a good example.
- <a href='https://github.com/tmeasday'></a> I think it's a good example.
Re the `onResultReceived` I would at least call out that it exists in the "method call lifecycle" step 4
- [ ] <a href='#crh-comment-Commit 9b9442b5f7e40a430b58bf7a4acdeaba8320b277 content/methods.md 12 366'></a> &nbsp;<b><a href='https://github.com/meteor/guide/commit/9b9442b5f7e40a430b58bf7a4acdeaba8320b277#commitcomment-14863628'>File: content/methods.md:L366 (9b9442b)</a></b>
- <a href='https://github.com/tmeasday'></a> Link to http://guide.meteor.com/collections.html#local-collections (I'll update the id later today)
- [x] <a href='#crh-comment-Pull 4fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f content/methods.md 54'></a> &nbsp;<b><a href='https://github.com/meteor/guide/pull/138#discussion_r46914612'>File: content/methods.md:L1-417</a></b>
- <a href='https://github.com/tmeasday'></a> "a client"? -- I mean server-server methods are a thing. But maybe it's better to just not mention it.
- <a href='https://github.com/stubailo'></a> Yeah I think if we mentioned it we'd need to go into detail and explain the use cases - some sort of microservice situation. But we don't really do that in the guide so let's leave it for now.
- [x] <a href='#crh-comment-Pull 4fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f content/methods.md 79'></a> &nbsp;<b><a href='https://github.com/meteor/guide/pull/138#discussion_r46914642'>File: content/methods.md:L1-417</a></b>
- <a href='https://github.com/tmeasday'></a> (on the client)
- <a href='https://github.com/tmeasday'></a> Ok, I misunderstood. Maybe "run the validation code by itself, without.."
- [x] <a href='#crh-comment-Pull 4fd199c0ebb8f0ae8fcab659dbddf7ad7ed5328f content/methods.md 410'></a> &nbsp;<b><a href='https://github.com/meteor/guide/pull/138#discussion_r46915044'>File: content/methods.md:L1-417</a></b>
- <a href='https://github.com/tmeasday'></a> Yeah, pretty much. `$inc`, `$push`, etc + any 3rd party side effects

<a href='https://www.codereviewhub.com/meteor/guide/pull/138?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/meteor/guide/pull/138?approve=1'></a>&nbsp;<a href='https://github.com/meteor/guide/pull/138'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by stubailo** [comment]
When is a good time to use `onResultReceived`?

**Comment by tmeasday** [comment]
The only time I ever used it was when repeatedly calling a method that touched the same collection, and wanting to get a callback for a progress bar.

If I recall correctly, what was happening is all the success callbacks for all the methods calls were firing at once because they were all waiting on each other's data.

**Comment by stubailo** [comment]
Hmmm.. is that worth mentioning? I can't really think of a simple, non-contrived example. Perhaps this is a feature people could look for in the docs.

**Comment by stubailo** [comment]
Speaking of this article, can you take a glance at the structure? It's a bit different than the outline, let me know what you think of the ordering.

File: content/methods.md [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+    Todos.update(todoId, {
+      $set: { text: newText }
+    });
+  }
+});
+```
+
+#### Calling
+
+This method is callable from the client and server using [`Meteor.call`](http://docs.meteor.com/#/full/meteor_call). Note that you should only use a Method in the case where some code needs to be callable from the client; if you just want to modularize code that is only going to be called from the server, use a regular JavaScript function, not a Method.
```
Comment (tmeasday) [inline content/methods.md]: "a client" -- I mean server-server methods are a thing. But maybe it's better to just not mention it.

File: content/methods.md [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+
+If the Method throws an error, you get that in the first argument of the callback. If the Method succeeds, you get the result in the second argument and the first argument `err` will be `undefined`. For more information about errors, see the section below about error handling.
+
+### Advanced Method boilerplate
+
+Meteor Methods have several features which aren't immediately obvious, but every complex app will need them at some point. These features were added incrementally over several years in a backwards-compatible fashion, so unlocking the full capabilities of Methods requires a good amount of boilerplate. In this article we will first show you all of the code you need to write for each feature, then the next section will talk about a Method wrapper package we have developed to make it easier.
+
+Here are some of the functionality an ideal Method would have:
+
+1. Run validation code without calling the Method
```
Comment (tmeasday) [inline content/methods.md]: (on the client)

Comment (tmeasday) [inline content/methods.md]: Ok, I misunderstood. Maybe "run the validation code by itself, without.."

File: content/methods.md [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+### `onResultReceived`
+
+XXX this doesn't even work with ValidatedMethod, and I don't know when you would use it
+
+### Method retries
+
+If you call a Method from the client, and the user's internet connection disconnects before the result is received, Meteor assumes that the Method didn't actually run. When the connection is re-established, the Method call will be sent again. This means that, in certain situations, Methods can be sent more than once. This should only happen very rarely, but in the case where an extra method call could have negative consequences it is worth putting in extra effort to ensure that Methods are idempotent - that is, calling them multiple times doesn't result in additional changes to the database.
+
+XXX this sounds sketchy... I would guess due to consistent ID generation this isn't an issue with inserts, perhaps only with methods that call $inc?
```
Comment (tmeasday) [inline content/methods.md]: Yeah, pretty much. `$inc`, `$push`, etc + any 3rd party side effects

**Comment by tmeasday** [comment]
It's looking really good. Definitely agree with the restructuring.

File: content/methods.md (hunk 1 above)
Comment (stubailo) [inline content/methods.md]: Yeah I think if we mentioned it we'd need to go into detail and explain the use cases - some sort of microservice situation. But we don't really do that in the guide so let's leave it for now.

File: content/methods.md (hunk 2 above)
Comment (stubailo) [inline content/methods.md]: :+1:

**Comment by stubailo** [comment]
Outstanding questions:

1. Should we avoid using `$inc` in Todos? I'm not sure how the denormalizer works.
2. What to do about `onResultReceived`?
3. How much detail should I write about loading data with Methods?

File: content/methods.md (hunk 3 above)
Comment (stubailo) [inline content/methods.md]: :+1:<a href='#crh-update-none'></a>

File: content/methods.md (hunk 1 above)
Comment (stubailo) [inline content/methods.md]: :+1:<a href='#crh-update-complete'></a>

**Comment by tmeasday** [comment]
1. Let's discuss this
2. I think drop it. It's in the docs, it's not common that you need it, and you can make passing reference to it in the spot where you explain how methods work.
3. I think you should mention using a local collection and dumping data in there from a method. Then you can add the downsides about optimistic UI and overfetching on secondary access

**Pull request merged**